pkg/scheduler/core/selectclusters/group_clusters.go

```diff
@@ -0,0 +1,275 @@
+	Clusters []ClusterDetailInfo
+}
+
+// RegionInfo indicate the region information
+type RegionInfo struct {
+	Name              string
+	Score             int64
+	AvailableReplicas int64
+
+	ProviderName string
+
+	// Zones under this region, sorted by zone.Score descending.
+	Zones []ZoneInfo
+	// Clusters under this region, sorted by cluster.Score descending.
+	Clusters []ClusterDetailInfo
+}
+
+// ZoneInfo indicate the zone information
+type ZoneInfo struct {
+	Name              string
+	Score             int64
+	AvailableReplicas int64
+
+	RegionName   string
+	ProviderName string
+
+	// Clusters under this zone, sorted by cluster.Score descending.
+	Clusters []ClusterDetailInfo
+}
+
+// ClusterDetailInfo indicate the cluster information
+type ClusterDetailInfo struct {
+	Name              string
+	Score             int64
+	AvailableReplicas int64
+
+	Cluster *clusterv1alpha1.Cluster
+}
+
+// GroupClustersWithScore groups cluster base provider/region/zone/cluster
+func GroupClustersWithScore(
+	clustersScore framework.ClusterScoreList,
+	placement *policyv1alpha1.Placement,
+	spec *workv1alpha2.ResourceBindingSpec,
+) *GroupClustersInfo {
+	if isTopologyIgnored(placement) {
+		return groupClustersIngoreTopology(clustersScore, spec)
+	}
+
+	return groupClustersBasedTopology(clustersScore, spec)
+}
+
+func groupClustersBasedTopology(
+	clustersScore framework.ClusterScoreList,
+	rbSpec *workv1alpha2.ResourceBindingSpec,
+) *GroupClustersInfo {
+	groupClustersInfo := &GroupClustersInfo{}
+	groupClustersInfo.generateClustersInfo(clustersScore, rbSpec)
+	groupClustersInfo.generateZoneInfo()
+	groupClustersInfo.generateRegionInfo()
+	groupClustersInfo.generateProviderInfo()
+
+	return groupClustersInfo
+}
+
+func groupClustersIngoreTopology(
+	clustersScore framework.ClusterScoreList,
+	rbSpec *workv1alpha2.ResourceBindingSpec,
+) *GroupClustersInfo {
+	groupClustersInfo := &GroupClustersInfo{}
+	groupClustersInfo.generateClustersInfo(clustersScore, rbSpec)
+
+	return groupClustersInfo
+}
+
+func (info *GroupClustersInfo) generateClustersInfo(clustersScore framework.ClusterScoreList, rbSpec *workv1alpha2.ResourceBindingSpec) {
+	var clusters []*clusterv1alpha1.Cluster
+	for _, clusterScore := range clustersScore {
+		clusterInfo := ClusterDetailInfo{}
+		clusterInfo.Name = clusterScore.Cluster.Name
+		clusterInfo.Score = clusterScore.Score
+		clusterInfo.Cluster = clusterScore.Cluster
+		info.Clusters = append(info.Clusters, clusterInfo)
+		clusters = append(clusters, clusterScore.Cluster)
+	}
+
+	clustersReplicas := calAvailableReplicas(clusters, rbSpec)
+	for i, clustersReplica := range clustersReplicas {
+		info.Clusters[i].AvailableReplicas = int64(clustersReplica.Replicas)
+	}
+
+	sortClusters(info.Clusters)
+}
+
+func (info *GroupClustersInfo) generateZoneInfo() {
+	zoneInfoMap := make(map[string]ZoneInfo)
+
+	for _, clusterInfo := range info.Clusters {
+		zone := clusterInfo.Cluster.Spec.Zone
```

Comment:
clusterInfo.Cluster.Spec.Zone may be empty.

Reply:
+1

Reply:
> clusterInfo.Cluster.Spec.Zone may be empty.
OK. added a condition **clusterInfo.Cluster.Spec.Zone != ""** and the provider/region also add the condition